### PR TITLE
Make manifest

### PR DESCRIPTION
**Pull request by amanj120**

### CODE REVIEW

File: platform/android/export/export.cpp [hunk 1]
```diff
@@ -778,6 +778,87 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+		String manifest_gles = _get_gles_tag();
+
+		Vector<String> perms;
+		_get_permissions(p_preset, p_give_internet, perms);
+		String manifest_permission;
+		for (int i = 0; i < perms.size(); i++) {
+			manifest_permission = manifest_permission.insert(0, "\t<uses-permission android:name=\"" + perms.get(i) + "\"/>\n");
+		}
+
+		String xr_metadata_text = "";
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Does everything work if this line looks like this?
`String xr_metadata_text;`

If so, consider simplifying here and below.

Note that you can delete these lines if you use the `manifest_text +=` approach.

File: platform/android/export/export.cpp [hunk 2]
```diff
@@ -778,6 +778,87 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+			int hand_tracking_index = p_preset->get("xr_features/hand_tracking"); // 0: none, 1: optional, 2: required
+			if (hand_tracking_index == 2) {
+				hand_tracking = "<uses-feature android:name=\"oculus.software.handtracking\" android:required=\"true\"/>\n";
+			} else if (hand_tracking_index > 0) {
+				hand_tracking = "<uses-feature android:name=\"oculus.software.handtracking\" android:required=\"false\"/>\n";
+			}
+		}
+
+		String package_name = p_preset->get("package/unique_name");
+		String instrumentation = "<instrumentation\n"
```
Comment (amanj120) [inline platform/android/export/export.cpp]: I tried moving the instrumentation class to the `app` module, but then it didn't build.

File: platform/android/export/export.cpp [hunk 3]
```diff
@@ -778,6 +778,87 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+				"    xmlns:tools=\"http://schemas.android.com/tools\">\n";
+
+		String manifest_screen_sizes = _get_screen_sizes_tag(p_preset);
+		String manifest_gles = _get_gles_tag();
+
+		Vector<String> perms;
+		_get_permissions(p_preset, p_give_internet, perms);
+		String manifest_permission;
+		for (int i = 0; i < perms.size(); i++) {
+			manifest_permission = manifest_permission.insert(0, "\t<uses-permission android:name=\"" + perms.get(i) + "\"/>\n");
```
Comment (thanwah) [inline platform/android/export/export.cpp]: The "\t" will cause the generated file to have a mix of tabs and spaces. Suggest replacing with the correct number of spaces.

File: platform/android/export/export.cpp [hunk 4]
```diff
@@ -778,6 +778,87 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+		}
+
+		String xr_metadata_text = "";
+		String fa_metadata_text = "";
+		String head_tracking = "";
+		String hand_tracking = "";
+		int xr_mode_index = p_preset->get("xr_features/xr_mode");
+		if (xr_mode_index == 1 /* XRMode.OVR */) {
+			String focus_awareness = bool_to_string(p_preset->get("xr_features/focus_awareness"));
+			xr_metadata_text = "<meta-data android:name=\"com.samsung.android.vr.application.mode\" android:value=\"vr_only\" />\n";
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Should any of these lines have an indent?

File: platform/android/export/export.cpp [hunk 5]
```diff
@@ -778,6 +778,87 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+			fa_metadata_text = "<meta-data android:name=\"com.oculus.vr.focusaware\" android:value=\"" + focus_awareness + "\"/>\n";
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Nit: on the previous line you used " />" but here use "/>". Please be consistent. (My personal preference is for the former.)

File: platform/android/export/export.cpp [hunk 6]
```diff
@@ -778,6 +778,87 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+			int dof_index = p_preset->get("xr_features/degrees_of_freedom"); // 0: none, 1: 3dof and 6dof, 2: 6dof
+			if (dof_index == 2) {
+				head_tracking = "<uses-feature android:name=\"android.hardware.vr.headtracking\" android:required=\"true\" android:version=\"1\"/>\n";
+			} else if (dof_index > 0) {
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Based on that comment above this should be the same as "dof_index == 1" ?

File: platform/android/export/export.cpp [hunk 7]
```diff
@@ -778,6 +778,87 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+				head_tracking = "<uses-feature android:name=\"android.hardware.vr.headtracking\" android:required=\"false\" android:version=\"1\"/>\n";
+			}
```
Comment (thanwah) [inline platform/android/export/export.cpp:803]: Do we not need the "<uses-feature " for this in the "dof_index == 0" or other cases?

File: platform/android/export/export.cpp [hunk 8]
```diff
@@ -778,6 +778,87 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+			int hand_tracking_index = p_preset->get("xr_features/hand_tracking"); // 0: none, 1: optional, 2: required
```
Comment (thanwah) [inline platform/android/export/export.cpp:804]: Same questions as above

File: platform/android/export/export.cpp (hunk 2 above)
Comment (thanwah) [inline platform/android/export/export.cpp]: Do these lines wrap better if you add a newline after `=` ?

File: platform/android/export/export.cpp [hunk 9]
```diff
@@ -778,6 +778,87 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 		}
 	}
 
+	void _write_tmp_manifest(const Ref<EditorExportPreset> &p_preset, bool p_give_internet, bool p_debug) {
+		String manifest_header =
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Alternative to consider:

		String manifest_text =
				"<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
				"<manifest xmlns:android=\"http://schemas.android.com/apk/res/android\"\n"
				"    xmlns:tools=\"http://schemas.android.com/tools\">\n";

		manifest_text += _get_screen_sizes_tag(p_preset);
		manifest_text += _get_gles_tag();

File: platform/android/export/export.cpp [hunk 10]
```diff
@@ -778,6 +778,87 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+		String package_name = p_preset->get("package/unique_name");
+		String instrumentation = "<instrumentation\n"
+								 "        tools:node=\"replace\"\n"
+								 "        android:name=\".GodotInstrumentation\"\n"
+								 "        android:icon=\"@mipmap/icon\"\n"
+								 "        android:label=\"@string/godot_project_name_string\"\n"
+								 "        android:targetPackage=\"" +
+								 package_name + "\" />\n";
+
+		String manifest_chunk_1 = manifest_header + manifest_screen_sizes + manifest_gles + manifest_permission + head_tracking + hand_tracking + instrumentation;
```
Comment (thanwah) [inline platform/android/export/export.cpp]: See above about maybe deleting this line

File: platform/android/export/export.cpp [hunk 11]
```diff
@@ -778,6 +778,87 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+		String manifest_activity = "<activity android:name=\"com.godot.game.GodotApp\" tools:replace=\"android:screenOrientation\" android:screenOrientation=\"landscape\">\n";
+		int orientation = p_preset->get("screen/orientation"); //default value is landscape
+		if (orientation == 1) {
+			manifest_activity = "<activity android:name=\"com.godot.game.GodotApp\" tools:replace=\"android:screenOrientation\" android:screenOrientation=\"portrait\">\n";
+		}
+
+		String manifest_chunk_2 = application_header + manifest_plugins + xr_metadata_text + manifest_activity + fa_metadata_text + "\n</activity>\n</application>\n</manifest>";
+		String manifest_text = manifest_chunk_1 + manifest_chunk_2;
+
+		String manifest_path = "res://android/build/src/release/AndroidManifest.xml";
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Please change to one of these two options:

`String manifest_path = p_debug ? "res://android/build/src/debug/AndroidManifest.xml" : "res://android/build/src/release/AndroidManifest.xml";`

```
String manifest_location = p_debug ? "debug" : "release";
String manifest_path = "res://android/build/src/" + manifest_location + "/AndroidManifest.xml";
```

File: platform/android/export/export.cpp [hunk 12]
```diff
@@ -2415,13 +2499,17 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 			String remote_plugins_binaries = get_plugins_binaries(BINARY_TYPE_REMOTE, enabled_plugins);
 			String custom_maven_repos = get_plugins_custom_maven_repos(enabled_plugins);
 			bool clean_build_required = is_clean_build_required(enabled_plugins);
+			String version_code = itos(p_preset->get("version/code"));
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Move these 2 lines right below "String package_name = get_package_name" to keep the ordering and organization consistent.

File: platform/android/export/gradle_export_util.h [hunk 13]
```diff
@@ -142,4 +142,33 @@ Error _create_project_name_strings_files(const Ref<EditorExportPreset> &p_preset
...
 }
 
+String bool_to_string(bool v) {
+	return v ? "true" : "false";
+}
+
+String _get_gles_tag() {
+	bool min_gles3 = ProjectSettings::get_singleton()->get("rendering/quality/driver/driver_name") == "GLES3" &&
+					 !ProjectSettings::get_singleton()->get("rendering/quality/driver/fallback_to_gles2");
+	String manifest_gles = "";
```
Comment (thanwah) [inline platform/android/export/gradle_export_util.h]: `return min_gles3 ? "<uses-feature tools:node=\"replace\" android:glEsVersion=\"0x00030000\" android:required=\"true\" />\n" : "";`

File: platform/android/java/app/config.gradle [hunk 14]
```diff
@@ -28,6 +28,22 @@ ext.getExportPackageName = { ->
...
+ext.getExportVersionCode = { ->
+    String versionCode = project.hasProperty("version_code") ? project.property("version_code") : "" //1 is the default version code
+    if (versionCode == null || versionCode.isEmpty()) {
+        versionCode = "1"
+    }
+    return Integer.parseInt(versionCode)
+}
+
+ext.getExportVersionName = { ->
+    String versionName = project.hasProperty("version_name") ? project.property("version_name") : "" //1 is the default version code
```
Comment (thanwah) [inline platform/android/java/app/config.gradle]: Incorrect copy-paste comment here. Go ahead and delete though.

File: platform/android/java/app/config.gradle [hunk 15]
```diff
@@ -28,6 +28,22 @@ ext.getExportPackageName = { ->
     return appId
 }
 
+ext.getExportVersionCode = { ->
+    String versionCode = project.hasProperty("version_code") ? project.property("version_code") : "" //1 is the default version code
```
Comment (thanwah) [inline platform/android/java/app/config.gradle]: Delete " //1 is the default version code"

File: platform/android/java/app/config.gradle [hunk 16]
```diff
@@ -28,6 +28,22 @@ ext.getExportPackageName = { ->
...
+    if (versionCode == null || versionCode.isEmpty()) {
```
Comment (thanwah) [inline platform/android/java/app/config.gradle:33]: Can you use ?: here?

File: platform/android/java/app/config.gradle [hunk 17]
```diff
@@ -28,6 +28,22 @@ ext.getExportPackageName = { ->
...
+        versionCode = "1"
+    }
+    return Integer.parseInt(versionCode)
+}
+
+ext.getExportVersionName = { ->
+    String versionName = project.hasProperty("version_name") ? project.property("version_name") : "" //1 is the default version code
+    if (versionName == null || versionName.isEmpty()) {
```
Comment (thanwah) [inline platform/android/java/app/config.gradle:41]: Can you use ?: here?

**Review by thanwah — CHANGES_REQUESTED** [review]

File: platform/android/java/app/config.gradle (hunk 16 above)
Comment (amanj120) [reply platform/android/java/app/config.gradle:33]: The function that this is based off of doesn't use ternary operators here. I think it's safer to follow [precedent](https://github.com/amanj120/godot/pull/17/files#diff-98ef5299ba59486356224f56e890e5f6R25)

Comment (thanwah) [reply platform/android/java/app/config.gradle:33]: Sure, that's fine.

File: platform/android/export/export.cpp [hunk 18]
```diff
@@ -778,6 +778,75 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+				"    xmlns:tools=\"http://schemas.android.com/tools\">\n";
+
+		manifest_text += _get_screen_sizes_tag(p_preset);
+		manifest_text += _get_gles_tag();
+
+		Vector<String> perms;
+		_get_permissions(p_preset, p_give_internet, perms);
+		for (int i = 0; i < perms.size(); i++) {
+			manifest_text +=
+					"    <uses-permission android:name=\"" + perms.get(i) + "\"/>\n";
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Can this be on one line or did clang-format do this unfortunate wrap? :(

File: platform/android/export/export.cpp [hunk 19]
```diff
@@ -778,6 +778,75 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+		String fa_metadata_text = "";
+		int xr_mode_index = p_preset->get("xr_features/xr_mode");
+		if (xr_mode_index == 1 /* XRMode.OVR */) {
+			String focus_awareness = bool_to_string(p_preset->get("xr_features/focus_awareness"));
+			xr_metadata_text = "<tools:node=\"replace\" meta-data android:name=\"com.samsung.android.vr.application.mode\" android:value=\"vr_only\" />\n";
+			fa_metadata_text = "<tools:node=\"replace\" meta-data android:name=\"com.oculus.vr.focusaware\" android:value=\"" + focus_awareness + "\"/>\n";
+			int dof_index = p_preset->get("xr_features/degrees_of_freedom"); // 0: none, 1: 3dof and 6dof, 2: 6dof
+			if (dof_index == 2) {
+				manifest_text += "<tools:node=\"replace\" uses-feature android:name=\"android.hardware.vr.headtracking\" android:required=\"true\" android:version=\"1\"/>\n";
+			} else if (dof_index == 1s) {
```
Comment (thanwah) [inline platform/android/export/export.cpp]: == 1

Also suggest moving this block before the == 2 block

File: platform/android/export/export.cpp [hunk 20]
```diff
@@ -778,6 +778,75 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+				manifest_text += "<tools:node=\"replace\" uses-feature android:name=\"android.hardware.vr.headtracking\" android:required=\"false\" android:version=\"1\"/>\n";
+			}
+			int hand_tracking_index = p_preset->get("xr_features/hand_tracking"); // 0: none, 1: optional, 2: required
+			if (hand_tracking_index == 2) {
+				manifest_text += "<tools:node=\"replace\" uses-feature android:name=\"oculus.software.handtracking\" android:required=\"true\"/>\n";
+			} else if (hand_tracking_index == 1) {
```
Comment (thanwah) [inline platform/android/export/export.cpp]: == 1 block first

File: platform/android/export/export.cpp [hunk 21]
```diff
@@ -778,6 +778,75 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+			manifest_text +=
+					"    <uses-permission android:name=\"" + perms.get(i) + "\"/>\n";
+		}
+
+		String xr_metadata_text = "";
+		String fa_metadata_text = "";
+		int xr_mode_index = p_preset->get("xr_features/xr_mode");
+		if (xr_mode_index == 1 /* XRMode.OVR */) {
+			String focus_awareness = bool_to_string(p_preset->get("xr_features/focus_awareness"));
+			xr_metadata_text = "<tools:node=\"replace\" meta-data android:name=\"com.samsung.android.vr.application.mode\" android:value=\"vr_only\" />\n";
```
Comment (thanwah) [inline platform/android/export/export.cpp]: `manifest_text +=` ?

File: platform/android/export/export.cpp [hunk 22]
```diff
@@ -778,6 +778,75 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+			fa_metadata_text = "<tools:node=\"replace\" meta-data android:name=\"com.oculus.vr.focusaware\" android:value=\"" + focus_awareness + "\"/>\n";
```
Comment (thanwah) [inline platform/android/export/export.cpp]: `manifest_text +=` ?

File: platform/android/export/export.cpp [hunk 23]
```diff
@@ -778,6 +778,75 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+				"    <application android:label=\"@string/godot_project_name_string\" "
+				"        android:allowBackup=\"false\" tools:ignore=\"GoogleAppIndexingWarning\" "
+				"        android:icon=\"@mipmap/icon\">)\n";
+
+		String plugins_names = get_plugins_names(get_enabled_plugins(p_preset));
+		if (!plugins_names.empty()) {
+			manifest_text += "<tools:node=\"replace\" meta-data android:name=\"plugins\" android:value=\"" + plugins_names + "\"/>\n";
+		}
+
+		String manifest_activity = "<activity android:name=\"com.godot.game.GodotApp\" tools:replace=\"android:screenOrientation\" android:screenOrientation=\"landscape\">\n";
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Rather than declare here, make this the else block or part of a ?:

**Review by thanwah — CHANGES_REQUESTED** [review]

File: platform/android/export/export.cpp (hunk 21 above)
Comment (amanj120) [reply platform/android/export/export.cpp]: The metadata tags aren't supposed to be in the manifest in that order.

File: platform/android/export/export.cpp [hunk 24]
```diff
@@ -778,6 +778,75 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+
+		Vector<String> perms;
+		_get_permissions(p_preset, p_give_internet, perms);
+		for (int i = 0; i < perms.size(); i++) {
+			manifest_text += vformat("    <uses-permission android:name=\"%s\" />\n", perms.get(i));
+		}
+
+		String xr_metadata_text = "";
+		String fa_metadata_text = "";
+		int xr_mode_index = p_preset->get("xr_features/xr_mode");
```
Comment (thanwah) [inline platform/android/export/export.cpp]: `bool uses_xr = p_preset->get("xr_features/xr_mode") == 1;`

File: platform/android/export/export.cpp [hunk 25]
```diff
@@ -778,6 +778,75 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+		if (xr_mode_index == 1 /* XRMode.OVR */) {
```
Comment (thanwah) [inline platform/android/export/export.cpp]: `if (uses_xr)`

File: platform/android/export/export.cpp [hunk 26]
```diff
@@ -778,6 +778,75 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+			String focus_awareness = bool_to_string(p_preset->get("xr_features/focus_awareness"));
+			xr_metadata_text = "        <meta-data tools:node=\"replace\" android:name=\"com.samsung.android.vr.application.mode\" android:value=\"vr_only\" />\n";
+			fa_metadata_text = vformat("            <meta-data tools:node=\"replace\" android:name=\"com.oculus.vr.focusaware\" android:value=\"%s\" />\n", focus_awareness);
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Move the 3 lines above here down where you need to add them within a `if (uses_xr)`

File: platform/android/export/export.cpp [hunk 27]
```diff
@@ -778,6 +778,75 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+				"    <application android:label=\"@string/godot_project_name_string\"\n"
+				"        android:allowBackup=\"false\" tools:ignore=\"GoogleAppIndexingWarning\"\n"
+				"        android:icon=\"@mipmap/icon\">)\n";
+
+		String plugins_names = get_plugins_names(get_enabled_plugins(p_preset));
+		if (!plugins_names.empty()) {
+			manifest_text += vformat("    <meta-data tools:node=\"replace\" android:name=\"plugins\" android:value=\"%s\" />\n", plugins_names);
+		}
+
+		String orientation = int(p_preset->get("screen/orientation")) == 1 ? "portrait" : "landscape";
```
Comment (thanwah) [inline platform/android/export/export.cpp]: is this int() a function not a cast?

**Review by thanwah — CHANGES_REQUESTED** [review]

File: platform/android/export/export.cpp [hunk 28]
```diff
@@ -778,6 +778,79 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+				"tools:replace=\"android:screenOrientation\" "
+				"android:screenOrientation=\"%s\">\n",
+				orientation);
+
+		if (uses_xr) {
+			String focus_awareness = bool_to_string(p_preset->get("xr_features/focus_awareness"));
+			manifest_text += vformat("            <meta-data tools:node=\"replace\" android:name=\"com.oculus.vr.focusaware\" android:value=\"%s\" />\n", focus_awareness);
+		}
+
+		manifest_text += "        </activity>\n    </application>\n</manifest>";
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Suggestion:

```
manifest_text += "        </activity>\n";
manifest_text += "    </application>\n";
manifest_text += "</manifest>";
```

**Review by thanwah — APPROVED** [review]

**Comment by amanj120** [comment]
PR's are up on godotengine:godot now.

**Pull request closed without merging**